src/ar_tx_replay_pool.erl

```diff
@@ -0,0 +1,290 @@
+	WeaveState = create_state(RecentTXs),
+	TXIDSet = sets:from_list(
+		[MempoolTX#tx.id || MempoolTX <- MempoolTXs]
+	),
+	Mempool = #mempool {
+		tx_id_set = TXIDSet
+	},
+	verify_tx(
+		general_verification,
+		TX,
+		Diff,
+		Height,
+		WalletList,
+		WeaveState,
+		Mempool
+	).
+
+%% @doc Verify the transactions are valid for the block taken into account
+%% the given current difficulty and height, the previous blocks' wallet list,
+%% and recent weave transactions.
+verify_block_txs(TXs, Diff, Height, WalletList, RecentTXs) ->
+	WeaveState = create_state(RecentTXs),
+	{VerifiedTXs, _, _} = apply_txs(
+		TXs,
+		Diff,
+		Height,
+		WalletList,
+		WeaveState,
+		#mempool {}
+	),
+	case length(VerifiedTXs) of
+		L when L == length(TXs) ->
+			valid;
+		_ ->
+			invalid
+	end.
+
+%% @doc Pick a list of transactions from the mempool to mine on.
+%% Transactions have to be valid when applied on top of each other taken
+%% into account the current height and diff, recent weave transactions, and
+%% the wallet list. The total data size of chosen transactions does not
+%% exceed the block size limit. Before a valid subset of transactions is chosen,
+%% transactions are sorted from biggest to smallest and then from oldest
+%% block anchors to newest.
+pick_txs_to_mine(RecentTXs, Height, Diff, WalletList, TXs) ->
+	case ar_fork:height_1_8() of
+		H when Height >= H ->
+			pick_txs_to_mine_post_1_8(RecentTXs, Height, Diff, WalletList, TXs);
+		_ ->
+			pick_txs_to_mine_pre_1_8(Height, Diff, WalletList, TXs)
+	end.
+
+%% @doc Choose transactions to keep in the mempool after a new block is
+%% accepted. Transactions are verified independently from each other
+%% taking into account the given difficulty and height of the new block,
+%% the new recent weave transactions, and the new wallet list.
+pick_txs_to_keep_in_mempool(RecentTXs, TXs, Diff, Height, WalletList) ->
+	WeaveState = create_state(RecentTXs),
+	Mempool = #mempool{},
+	lists:filter(
+		fun(TX) ->
+			case verify_tx(
+				general_verification,
+				TX,
+				Diff,
+				Height,
+				WalletList,
+				WeaveState,
+				Mempool
+			) of
+				{valid, _, _} ->
+					true;
+				{invalid, _} ->
+					false
+			end
+		end,
+		TXs
+	).
+
+%% PRIVATE
+
+create_state(RecentTXs) ->
+	MaxDepthTXs = lists:sublist(RecentTXs, ?MAX_TX_ANCHOR_DEPTH),
+	{BHL, Map} = lists:foldr(
+		fun({BH, TXIDs}, {BHL, Map}) ->
+			{[BH | BHL], maps:put(BH, sets:from_list(TXIDs), Map)}
+		end,
+		{[], maps:new()},
+		MaxDepthTXs
+	),
+	#state {
+		weave_map = Map,
+		bhl = BHL
+	}.
+
+verify_tx(general_verification, TX, Diff, Height, FloatingWalletList, WeaveState, Mempool) ->
+	case ar_tx:verify(TX, Diff, Height, FloatingWalletList) of
+		true ->
+			verify_tx(last_tx_in_mempool, TX, Diff, Height, FloatingWalletList, WeaveState, Mempool);
+		false ->
+			{invalid, tx_verification_failed}
+	end;
+verify_tx(last_tx_in_mempool, TX, Diff, Height, FloatingWalletList, WeaveState, Mempool) ->
```

Comment:
How about refactoring this clause to:
```erlang
verify_tx(last_tx_in_mempool, TX, Diff, Height, FloatingWalletList, WeaveState, Mempool) ->
	ShouldContinue = case ar_fork:height_1_8() of
		H when Height >= H ->
			%% Only verify after fork 1.8 otherwise it causes a soft fork
			%% since current nodes can accept blocks with a chain of last_tx
			%% references. The check would still fail on edge pre 1.8 since
			%% TX is validated against a previous blocks' wallet list then.
			case sets:is_element(TX#tx.last_tx, Mempool#mempool.tx_id_set) of
				true ->
					{invalid, last_tx_in_mempool};
				false ->
					continue
			end;
		_ ->
			continue
	end,
	case ShouldContinue of
		continue ->
			verify_tx(
				last_tx,
				TX,
				Diff,
				Height,
				FloatingWalletList,
				WeaveState,
				Mempool
			);
		{invalid, Reason} ->
			{invalid, Reason}
	end;